capture/packetbuffer.go

```diff
@@ -13,84 +13,111 @@ var (
 	ErrPacketsFull = errors.New("PacketBuffer out of space for more packets")
 )
 
-// PacketBuffer stores packets in a compact format
-type PacketBuffer struct {
-	numPackets int
-	cis        []gopacket.CaptureInfo
+type BlockBuffer struct {
 	// stores where the nth packet ends in data
 	offsets []int
 	// packet n is located at data[offsets[n-1]:offsets[n]]
 	data []byte
 }
 
+func NewBlockBuffer(maxBlocks, maxBytes int) BlockBuffer {
+	return BlockBuffer{
+		offsets: make([]int, 0, maxBlocks),
+		data:    make([]byte, 0, maxBytes),
+	}
+}
+
+// PacketBuffer stores packets in a compact format
+type PacketBuffer struct {
+	BlockBuffer
+	cis []gopacket.CaptureInfo
+}
+
 // NewPacketBuffer creates a PacketBuffer with the specified limits.
 func NewPacketBuffer(maxPackets, maxBytes int) *PacketBuffer {
 	return &PacketBuffer{
-		cis:     make([]gopacket.CaptureInfo, maxPackets),
-		offsets: make([]int, maxPackets),
-		data:    make([]byte, maxBytes),
+		BlockBuffer: NewBlockBuffer(maxPackets, maxBytes),
+		cis:         make([]gopacket.CaptureInfo, maxPackets),
 	}
 }
 
-func (pb *PacketBuffer) bytesStored() int {
-	if pb.numPackets == 0 {
-		return 0
+// BytesRemaining returns the number of additional bytes this PacketBuffer can hold.
+func (b *BlockBuffer) BytesRemaining() int {
+	return cap(b.data) - len(b.data)
+}
+
+func (b *BlockBuffer) Block(n int) []byte {
+	start := 0
+	if n > 0 {
+		start = b.offsets[n-1]
 	}
-	return pb.offsets[pb.numPackets-1]
+	end := b.offsets[n]
+	return b.data[start:end]
 }
 
-// BytesRemaining returns the number of additional bytes this PacketBuffer can hold.
-func (pb *PacketBuffer) BytesRemaining() int {
-	return len(pb.data) - pb.bytesStored()
+func (b *BlockBuffer) BlockLen() int {
+	return len(b.offsets)
+}
+
+func (b *BlockBuffer) Clear() {
+	b.offsets = b.offsets[:0]
+	b.data = b.data[:0]
 }
 
 // Append adds a packet to the PacketBuffer.
 // Makes a copy of pd.Data, which may be modified after Append returns.
 // Returns ErrBytesFull or ErrPacketsFull if there is insufficient space.
-func (pb *PacketBuffer) Append(pd PacketData) error {
-	if pb.numPackets >= len(pb.offsets) {
+func (b *BlockBuffer) Append(data []byte) error {
+	if cap(b.offsets) == 0 {
+		panic("attempted to Append to uninitialized BlockBuffer")
+	}
+	if len(b.offsets) >= cap(b.offsets) {
 		return ErrPacketsFull
 	}
-	if pb.bytesStored()+pd.Info.CaptureLength > len(pb.data) {
+	start := len(b.data)
+	end := len(b.data) + len(data)
+	if end > cap(b.data) {
 		return ErrBytesFull
 	}
 
-	pb.cis[pb.numPackets] = pd.Info
-	pb.offsets[pb.numPackets] = pb.bytesStored() + pd.Info.CaptureLength
-	copy(pb.data[pb.bytesStored():], pd.Data)
-	pb.numPackets++
+	b.data = b.data[:end]
```

Comment:
ok, so this increases the "length" of a slice in golang? (since we've checked that the addition of "data" will be within the slice's "capacity")

if we did the copy below without this line, would we get the same result?

Reply:
also, would it work to use append() here? i think so, as we won't grow the capacity of the slice with append(), since we've already checked that condition above

Reply:
Without the reslicing, we would not get the same result. copy() copies until the end of either the source or the destination and does not (can not) modify the slice bounds.

We could use append(). It might be very fractionally slower since append would repeat the checks against capacity, but may be more readable.

Note that before this patch, we didn't actually re-slice b.data, but the tradeoff is more complex code in BytesRemaining() to determine how much of the b.data is actually in use.